### PR TITLE
Fix some Composite Compliance problems with binary_cross_entropy backward

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#70198 Fix some Composite Compliance problems with binary_cross_entropy backward**

This PR fixes composite compliance problems with:
- binary_cross_entropy's backward formula
- binary_cross_entropy_with_logits's backward formula
- binary_cross_entropy's double backward formula

It does so by adding checks for areAnyTensorSubclassLike.

Test Plan:
- I tested everything with functorch.
- We are going to do https://github.com/pytorch/pytorch/issues/69530 in
the future so we have a way of testing this in core. I need the
binary_cross_entropy ones for something right now and didn't want to
wait until we come up with a solution for #69530.

Differential Revision: [D33246995](https://our.internmc.facebook.com/intern/diff/D33246995)

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]
LGTM

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33246995).

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -1361,10 +1361,21 @@ Tensor binary_cross_entropy_target_backward(
   const Tensor& target,
   const c10::optional<Tensor>& weight,
   int64_t reduction) {
-  auto grad_target = (1. - self).log_().sub_(self.log()).mul_(grad);
+  auto grad_target = (1. - self).log_().sub_(self.log());
+  if (!areAnyTensorSubclassLike({grad})) {
+    // fast-path
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1366]: I am not sure such comment is useful.
If we want to explain why, we should have a note explaining when inplace are allowed and not right?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -1378,13 +1389,22 @@ Tensor binary_cross_entropy_target_backward(
 Tensor binary_cross_entropy_with_logits_target_backward(const Tensor& grad_output, const Tensor& self, const Tensor& target, const c10::optional<Tensor>& weight, const c10::optional<Tensor>& pos_weight, int64_t reduction) {
   Tensor grad_target;
   if (isDefined(pos_weight)) {
-    grad_target = (1. - self.sigmoid()).log_().sub_(pos_weight->mul(self.sigmoid().log_())).mul_(grad_output);
+    if (!areAnyTensorSubclassLike({*pos_weight, grad_output})) {
+      // fast-path
+      grad_target = (1. - self.sigmoid()).log_().sub_(pos_weight->mul(self.sigmoid().log_())).mul_(grad_output);
+    } else {
+      grad_target = (1. - self.sigmoid()).log_().sub(pos_weight->mul(self.sigmoid().log_())).mul(grad_output);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1396]: the mul with self.sigmoid() could be inplace no?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (zou3519) [reply torch/csrc/autograd/FunctionsManual.cpp:1366]: Sure, I can remove the comments

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:1396]: Ho that op is applied to pos_weight directly. Sorry with the line break I though it was an intermediary. We don't want to modify pos_weight inplace so we shouldn't indeed.

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33246995).
